### ISSUE TITLE
[rl] PPO/PPG: `max_grad_norm` is dead and `clip_grad` defaults to None -> no clipping, docs claim otherwise

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#183]
> **Rewritten 2026-07-17** after re-triage against the current tree. **Confirmed, and the real situation is worse than reported** — but the docs analysis was misattributed and PPG was omitted. Original text is in the edit history. See "Triage record" below.

## The bug

`PpoTrainingConfig` exposes `max_grad_norm` (`ppo_config.rs:52`, default `0.5` at `:131`, validated `positive` at `:162`, builder setter at `:246-248`) and **nothing ever reads it**. A workspace-wide search finds it only in that declaration/default/validate/setter cluster. It is a genuinely dead field.

## Worse: default PPO does no gradient clipping at all

There *is* a functional clipping knob — `clip_grad` (`ppo_config.rs:133`), wired into the optimizers at `ppo_agent.rs:247-259`. **It defaults to `None`.**

So there are two fields: the advertised one is dead, and the functional one is off by default. A user reading `max_grad_norm: 0.5` in the config reasonably concludes clipping is on. It is not.

## PPG has the identical defect — omitted from the original

`ppg_config.rs:28` — `pub ppo: PpoTrainingConfig` embeds it verbatim; `:78` delegates validation. `ppg_agent.rs:259-268` builds optimizers from `config.ppo.clip_grad` and never reads `max_grad_norm`. Any fix must cover both.

## The false claims to correct

- `README.md:152` — lists `max_grad_norm | 0.5 | Huang et al. #11` as an effective default. **False** (and it cites #11, not #10).
- `ppo_config.rs:49-52` — "applied to each `loss.backward()` result". **False.**
- `ppo_config.rs:58-61` — "applied manually in the agent loop". **False.**
- `ppo/README.md:44` — attributes CleanRL #10 to `clip_grad` and calls it **"Global"** gradient clipping. **Also inaccurate** — see below.

## Fix: delete the field, correct the docs

Remove `max_grad_norm`, its validator, and its setter from `PpoTrainingConfig`; fix the four doc sites above; do both for PPG.

**"Just wire `max_grad_norm` into the optimizer" is not a valid fix.** Burn's `Norm` clipping is **per-tensor, not global**: `burn-optim-0.21.0/src/grad_clipping/base.rs:72-91` takes the `l2_norm` of **one** gradient tensor, and `SimpleOptimizerMapper` (`optim/simple/adaptor.rs:83-85`) applies it per-parameter. CleanRL detail #10 clips the **global** norm across the entire flattened parameter vector. These are materially different operations — the per-tensor form neither bounds the global norm nor preserves the gradient's direction.

Implementing true global-norm clipping is a separate feature (Burn has no built-in; it needs a reduction across `GradientsParams` before `step_with`). Filed as **#328**. This issue should land regardless of whether #328 is ever built — the docs must stop claiming a feature that does not exist.

---

## Triage record

| Original claim | Verdict |
|---|---|
| `PpoTrainingConfig` exposes `max_grad_norm` | **CONFIRMED** (`ppo_config.rs:52`) |
| `update()` never applies it | **CONFIRMED** — genuinely dead field |
| Docs claim CleanRL detail **#10** | **PARTIAL / misattributed** — `ppo/README.md:44` attributes #10 to `clip_grad` and is *honest* about the field split. The false claims are at `README.md:152` (which cites **#11**), `ppo_config.rs:49-52`, and `ppo_config.rs:58-61`. |
| Scope: `ppo_agent.rs` only | **REFUTED** — PPG has the same dead field (`ppg_config.rs:28`) and is not named. Same omission pattern as #167. |
| Implied fix: wire it into the optimizer | **REFUTED** — Burn's `Norm` is per-tensor; it would not reproduce detail #10 |

**Already known internally:** `rlevo-examples/examples/classic/report_ppo_cartpole_with_client.rs:264-265` comments that `max_grad_norm` is "a separate, currently inert config field — `clip_grad` is the knob that actually clips." The inertness was documented in an example while the config kept advertising the opposite.

## Related

- **#328** — implement true global-norm clipping (or decide not to)

**Comment by anthonytorlucci** [comment]
Re-verified against the working tree with the #170 fix applied (uncommitted). **This issue is unaffected by #170** — its hunks begin at `ppo_agent.rs:379` / `ppg_agent.rs:438`, below the optimizer construction at `:247-259` / `:259-268`; it touched no `*_config.rs`; and its `ppo/README.md` edit is confined to lines 56-70, not line 44. The two can land in either order.

Commenting only to add a **fifth doc site** the fix list omits. The example this issue cites as internal evidence carries its own false claim two lines above the part quoted:

```
report_ppo_cartpole_with_client.rs:262
    // Global grad-norm clip of 0.5 (CleanRL default). The optimizer wraps
    // both policy and value grads with this; without it, PPO can take
    // destabilizing steps. NOTE: `max_grad_norm` is a separate, currently
    // inert config field — `clip_grad` is the knob that actually clips.
```

It is honest about *which field* clips (`:264-265`, correctly) and wrong about *what the clipping does* — calling Burn's per-tensor `Norm` a "Global grad-norm" clip is the exact conflation this issue spends a paragraph refuting. Worth adding `:262` alongside the four sites already listed.

While there, I confirmed the Burn analysis from vendored source rather than taking it on faith, since it decides which fixes are possible: `burn-optim-0.21.0/src/grad_clipping/base.rs:72-91` takes `l2_norm` of a single `Tensor<B, D>`, and `SimpleOptimizerMapper::map_float` (`optim/simple/adaptor.rs:178`) is driven by `module.map(...)` at `:83-90` — per-parameter. `rg -i "global.?norm|total_norm|clip_grad_norm"` over `burn-optim/src` returns nothing, and `GradientsParams` exposes only `get`/`remove`/`register` by `ParamId` plus `len` (`optim/grads.rs:71-104`) with no iterator. So global-norm clipping needs a module visitor to enumerate params first. The issue's refutation of "just wire it into the optimizer" holds.
